src/qt_gui/keyboardcontrolswindow.cpp

```diff
@@ -0,0 +1,620 @@
+    ui->CrossKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Cross_Key]));
+    ui->R2KeySequenceEdit->setKeySequence(convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::R2_Key]));
+    ui->CircleKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Circle_Key]));
+    ui->R1KeySequenceEdit->setKeySequence(convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::R1_Key]));
+    ui->SquareKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Square_Key]));
+    ui->TriangleKeySequenceEdit->setKeySequence(
+        convertSDLKeyToQt(m_reverseKeysMap[KeysMapping::Triangle_Key]));
+
+    QObject::connect(ui->applyButton, &QPushButton::clicked,
+                     [this]() { validateAndSaveKeyBindings(); });
+
+    QObject::connect(ui->cancelButton, &QPushButton::clicked, [this]() { this->close(); });
+
+    QList<QKeySequenceEdit*> keySequenceEdits = {
+        ui->StartKeySequenceEdit,        ui->SelectKeySequenceEdit,
+        ui->LAnalogDownkeySequenceEdit,  ui->LAnalogLeftkeySequenceEdit,
+        ui->LAnalogUpkeySequenceEdit,    ui->LAnalogRightkeySequenceEdit,
+        ui->PSkeySequenceEdit,           ui->RAnalogDownkeySequenceEdit,
+        ui->RAnalogLeftkeySequenceEdit,  ui->RAnalogUpkeySequenceEdit,
+        ui->RAnalogRightkeySequenceEdit, ui->DPadLeftkeySequenceEdit,
+        ui->DPadRightkeySequenceEdit,    ui->DPadUpkeySequenceEdit,
+        ui->DPadDownkeySequenceEdit,     ui->L2keySequenceEdit,
+        ui->L1keySequenceEdit,           ui->CrossKeySequenceEdit,
+        ui->R2KeySequenceEdit,           ui->CircleKeySequenceEdit,
+        ui->R1KeySequenceEdit,           ui->SquareKeySequenceEdit,
+        ui->TriangleKeySequenceEdit};
+
+    for (auto edit : keySequenceEdits) {
+        edit->setStyleSheet("QLineEdit { qproperty-alignment: AlignCenter; }");
+    }
+}
+
+KeyboardControlsWindow::~KeyboardControlsWindow() {
+    delete ui;
+}
+
+std::map<Uint32, KeysMapping> KeyboardControlsWindow::getKeysMapping() {
+    return m_keysMap;
+}
+
+void KeyboardControlsWindow::validateAndSaveKeyBindings() {
+    int nOfUnconfiguredButtons = 0;
+    for (auto& keyEdit : m_listOfKeySequenceEdits) {
+        auto keySequence = keyEdit->keySequence();
+
+        if (keySequence == QKeySequence()) {
+            keyEdit->setStyleSheet("background-color: red; qproperty-alignment: AlignCenter;");
+            QTimer::singleShot(2000, keyEdit, [keyEdit]() {
+                keyEdit->setStyleSheet("qproperty-alignment: AlignCenter;"); // Reset to default
+            });
+
+            ++nOfUnconfiguredButtons;
+        }
+    }
+
+    if (nOfUnconfiguredButtons > 0) {
+        showError("Some of the buttons were not configured");
+        return;
+    }
+
+    m_keysMap.clear();
+    m_reverseKeysMap.clear();
+
+    m_keysMap.emplace(convertQtKeyToSDL(ui->LAnalogDownkeySequenceEdit->keySequence()[0].key()),
```

Comment:
Lots of duplicated code here again that can be reduced with a loop and that QList

Reply:
unfortunatelly here I can't use that QList... I have to explicitly map a QKeySequenceEdit's value to an enum, if I just use the list I will not know which one is which